File: app/src/main/java/com/bas080/autosleepdroid/EventLogger.java
package com.bas080.autosleepdroid;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Typeface;
import android.os.Handler;
import android.os.Looper;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class EventLogger {
    public static final int LEVEL_LOW = 0;
    public static final int LEVEL_NORMAL = 1;
    public static final int LEVEL_HIGH = 2;

    private static final String PREF_NAME = "event_logger";
    private static final String KEY_LOGS = "logs";
    private static final int MAX_LOGS = 500;
    private static final List<String> events = new ArrayList<>();
    private static boolean loaded = false;
    private static Listener listener;
    private static Context appContext;

    public interface Listener {
        void onEventLogged(String event);
    }

    public static synchronized void setListener(Listener l) {
        listener = l;
    }

    public static synchronized void log(Context context, int level, String message) {
        if (context != null) {
            appContext = context.getApplicationContext();
        }
        Context targetContext = context != null ? context : appContext;
        ensureLoaded(targetContext);

        String timestamp = new SimpleDateFormat("M/d HH:mm:ss", Locale.US).format(new Date());
        char marker;
        switch (level) {
            case LEVEL_LOW:
                marker = '\u0000';
                break;
            case LEVEL_HIGH:
                marker = '\u0002';
                break;
            case LEVEL_NORMAL:
            default:
                marker = '\u0001';
                break;
        }

        String line = timestamp + " " + marker + message;

        events.add(line);
        if (events.size() > MAX_LOGS) {
            events.remove(0);
        }

        if (targetContext != null) {
            persistLogs(targetContext.getApplicationContext());
        }

        final Listener currentListener = listener;
        if (currentListener != null) {
            if (Looper.myLooper() == Looper.getMainLooper()) {
                currentListener.onEventLogged(line);
            } else {
                new Handler(Looper.getMainLooper()).post(() -> currentListener.onEventLogged(line));
            }
        }
    }

    public static synchronized void log(Context context, String message) {
        log(context, LEVEL_NORMAL, message);
    }

    public static synchronized void log(int level, String message) {
        log(null, level, message);
    }

    public static synchronized void log(String message) {
        log(null, LEVEL_NORMAL, message);
    }

    public static synchronized List<String> getEvents(Context context) {
        ensureLoaded(context);
        return Collections.unmodifiableList(new ArrayList<>(events));
    }

    public static boolean isDarkMode(Context context) {
        if (context == null) {
            context = appContext;
        }
        if (context == null || context.getResources() == null || context.getResources().getConfiguration() == null) {
            return false;
        }
        int currentNightMode = context.getResources().getConfiguration().uiMode & android.content.res.Configuration.UI_MODE_NIGHT_MASK;
        return currentNightMode == android.content.res.Configuration.UI_MODE_NIGHT_YES;
    }

    public static CharSequence formatColoredEvent(String line) {
        return formatColoredEvent(null, line);
    }

    public static CharSequence formatColoredEvent(Context context, String line) {
        if (line == null || line.isEmpty()) {
            return "";
        }

        boolean darkMode = isDarkMode(context);

        int spaceIdx = line.indexOf(' ');
        int secondSpaceIdx = spaceIdx != -1 ? line.indexOf(' ', spaceIdx + 1) : -1;
        int timestampEnd = secondSpaceIdx != -1 ? secondSpaceIdx : (spaceIdx != -1 ? spaceIdx : 0);

        int level = LEVEL_NORMAL;
        String timestamp = timestampEnd > 0 ? line.substring(0, timestampEnd) : "";
        String rawMessage = timestampEnd < line.length() ? line.substring(timestampEnd) : "";

        if (rawMessage.contains("\u0000")) {
            level = LEVEL_LOW;
            rawMessage = rawMessage.replace("\u0000", "");
        } else if (rawMessage.contains("\u0002")) {
            level = LEVEL_HIGH;
            rawMessage = rawMessage.replace("\u0002", "");
        } else if (rawMessage.contains("\u0001")) {
            level = LEVEL_NORMAL;
            rawMessage = rawMessage.replace("\u0001", "");
        }

        String displayString = timestamp.isEmpty() ? rawMessage.trim() : (timestamp + rawMessage);
        SpannableString spannable = new SpannableString(displayString);

        int timestampColor = darkMode ? 0xFF888888 : 0xFF999999;
        if (!timestamp.isEmpty() && timestamp.length() <= displayString.length()) {
            spannable.setSpan(new ForegroundColorSpan(timestampColor), 0, timestamp.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        }

        int messageStart = timestamp.isEmpty() ? 0 : timestamp.length();

        int textColor;
        boolean isBold = false;

        if (darkMode) {
            if (level == LEVEL_LOW) {
                textColor = 0xFFAAAAAA;
            } else if (level == LEVEL_HIGH) {
                textColor = 0xFFFFFFFF;
                isBold = true;
            } else {
                textColor = 0xFFDDDDDD;
            }
        } else {
            if (level == LEVEL_LOW) {
                textColor = 0xFF888888;
            } else if (level == LEVEL_HIGH) {
                textColor = 0xFF000000;
                isBold = true;
            } else {
                textColor = 0xFF444444;
            }
        }

        if (messageStart < displayString.length()) {
            spannable.setSpan(new ForegroundColorSpan(textColor), messageStart, displayString.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            if (isBold) {
                spannable.setSpan(new StyleSpan(Typeface.BOLD), messageStart, displayString.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
        }

        return spannable;
    }

    public static synchronized void clear(Context context) {
        events.clear();
        appContext = null;
        loaded = false;
        if (context != null) {
            SharedPreferences prefs = context.getApplicationContext().getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
            prefs.edit().remove(KEY_LOGS).apply();
        }
    }

    private static void ensureLoaded(Context context) {
        if (!loaded && context != null) {
            SharedPreferences prefs = context.getApplicationContext().getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
            String raw = prefs.getString(KEY_LOGS, "");
            if (raw != null && !raw.isEmpty()) {
                String[] lines = raw.split("\n");
                for (String line : lines) {
                    if (!line.trim().isEmpty()) {
                        events.add(line);
                    }
                }
            }
            loaded = true;
        }
    }

    private static void persistLogs(Context context) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < events.size(); i++) {
            if (i > 0) sb.append("\n");
            sb.append(events.get(i));
        }
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        prefs.edit().putString(KEY_LOGS, sb.toString()).apply();
    }
}

File: app/src/main/java/com/bas080/autosleepdroid/EventLogger.kt
package com.bas080.autosleepdroid

import android.content.Context
import android.graphics.Typeface
import android.os.Handler
import android.os.Looper
import android.text.SpannableString
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import java.io.File
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Collections
import java.util.Date
import java.util.Locale

object EventLogger {
    const val LEVEL_LOW: Int = 0
    const val LEVEL_NORMAL: Int = 1
    const val LEVEL_HIGH: Int = 2

    private const val LOG_FILE_NAME = "event_logs.txt"
    private const val PREF_NAME = "event_logger"
    private const val KEY_LOGS = "logs"
    private const val MAX_LOGS = 500
    private val events = ArrayList<String>()
    private var loaded = false
    @Volatile
    private var listener: Listener? = null
    @Volatile
    private var appContext: Context? = null

    fun interface Listener {
        fun onEventLogged(event: String)
    }

    @Synchronized
    fun setListener(l: Listener?) {
        listener = l
    }

    @Synchronized
    fun log(context: Context?, level: Int, message: String) {
        if (context != null) {
            appContext = context.applicationContext
        }
        val targetContext = context ?: appContext
        ensureLoaded(targetContext)

        val timestamp = SimpleDateFormat("M/d HH:mm:ss", Locale.US).format(Date())
        val marker = when (level) {
            LEVEL_LOW -> '\u0000'
            LEVEL_HIGH -> '\u0002'
            else -> '\u0001'
        }

        val line = "$timestamp $marker$message"

        events.add(line)
        val exceeded = events.size > MAX_LOGS
        if (exceeded) {
            events.removeAt(0)
        }

        if (targetContext != null) {
            val appCtx = targetContext.applicationContext
            if (exceeded) {
                rewriteLogs(appCtx)
            } else {
                appendLogLine(appCtx, line)
            }
        }

        val currentListener = listener
        if (currentListener != null) {
            if (Looper.myLooper() == Looper.getMainLooper()) {
                currentListener.onEventLogged(line)
            } else {
                Handler(Looper.getMainLooper()).post { currentListener.onEventLogged(line) }
            }
        }
    }

    @Synchronized
    fun log(context: Context?, message: String) {
        log(context, LEVEL_NORMAL, message)
    }

    @Synchronized
    fun log(level: Int, message: String) {
        log(null, level, message)
    }

    @Synchronized
    fun log(message: String) {
        log(null, LEVEL_NORMAL, message)
    }

    @Synchronized
    fun getEvents(context: Context?): List<String> {
        ensureLoaded(context)
        return Collections.unmodifiableList(ArrayList(events))
    }

    fun isDarkMode(context: Context?): Boolean {
        val ctx = context ?: appContext ?: return false
        val resources = ctx.resources ?: return false
        val config = resources.configuration ?: return false
        val currentNightMode = config.uiMode and android.content.res.Configuration.UI_MODE_NIGHT_MASK
        return currentNightMode == android.content.res.Configuration.UI_MODE_NIGHT_YES
    }

    fun formatColoredEvent(line: String?): CharSequence {
        return formatColoredEvent(null, line)
    }

    fun formatColoredEvent(context: Context?, line: String?): CharSequence {
        if (line.isNullOrEmpty()) {
            return ""
        }

        val darkMode = isDarkMode(context)

        val spaceIdx = line.indexOf(' ')
        val secondSpaceIdx = if (spaceIdx != -1) line.indexOf(' ', spaceIdx + 1) else -1
        val timestampEnd = if (secondSpaceIdx != -1) secondSpaceIdx else (if (spaceIdx != -1) spaceIdx else 0)

        var level = LEVEL_NORMAL
        val timestamp = if (timestampEnd > 0) line.substring(0, timestampEnd) else ""
        var rawMessage = if (timestampEnd < line.length) line.substring(timestampEnd) else ""

        if (rawMessage.contains("\u0000")) {
            level = LEVEL_LOW
            rawMessage = rawMessage.replace("\u0000", "")
        } else if (rawMessage.contains("\u0002")) {
            level = LEVEL_HIGH
            rawMessage = rawMessage.replace("\u0002", "")
        } else if (rawMessage.contains("\u0001")) {
            level = LEVEL_NORMAL
            rawMessage = rawMessage.replace("\u0001", "")
        }

        val displayString = if (timestamp.isEmpty()) rawMessage.trim() else (timestamp + rawMessage)
        val spannable = SpannableString(displayString)

        val timestampColor = if (darkMode) -0x777778 else -0x666667
        if (timestamp.isNotEmpty() && timestamp.length <= displayString.length) {
            spannable.setSpan(ForegroundColorSpan(timestampColor), 0, timestamp.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        }

        val messageStart = if (timestamp.isEmpty()) 0 else timestamp.length

        val textColor: Int
        var isBold = false

        if (darkMode) {
            if (level == LEVEL_LOW) {
                textColor = -0x555556
            } else if (level == LEVEL_HIGH) {
                textColor = -0x1
                isBold = true
            } else {
                textColor = -0x222223
            }
        } else {
            if (level == LEVEL_LOW) {
                textColor = -0x777778
            } else if (level == LEVEL_HIGH) {
                textColor = -0x1000000
                isBold = true
            } else {
                textColor = -0xbbbbbc
            }
        }

        if (messageStart < displayString.length) {
            spannable.setSpan(ForegroundColorSpan(textColor), messageStart, displayString.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            if (isBold) {
                spannable.setSpan(StyleSpan(Typeface.BOLD), messageStart, displayString.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        }

        return spannable
    }

    @Synchronized
    fun clear(context: Context?) {
        events.clear()
        val targetContext = context ?: appContext
        if (targetContext != null) {
            val appCtx = targetContext.applicationContext
            val file = getLogFile(appCtx)
            if (file.exists()) {
                file.delete()
            }
            val prefs = appCtx.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            prefs.edit().remove(KEY_LOGS).commit()
        }
        if (context != null) {
            appContext = context.applicationContext
        } else {
            appContext = null
        }
        loaded = false
    }

    private fun getLogFile(context: Context): File {
        return File(context.filesDir, LOG_FILE_NAME)
    }

    private fun ensureLoaded(context: Context?) {
        if (!loaded && context != null) {
            val appCtx = context.applicationContext
            val file = getLogFile(appCtx)
            if (file.exists()) {
                file.useLines { lines ->
                    for (line in lines) {
                        if (line.trim().isNotEmpty()) {
                            events.add(line)
                        }
                    }
                }
            } else {
                val prefs = appCtx.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
                val raw = prefs.getString(KEY_LOGS, null)
                if (!raw.isNullOrEmpty()) {
                    val lines = raw.split("\n")
                    for (line in lines) {
                        if (line.trim().isNotEmpty()) {
                            events.add(line)
                        }
                    }
                    prefs.edit().remove(KEY_LOGS).commit()
                }
            }

            while (events.size > MAX_LOGS) {
                events.removeAt(0)
            }

            if (!file.exists() && events.isNotEmpty()) {
                rewriteLogs(appCtx)
            }

            loaded = true
        }
    }

    private fun appendLogLine(context: Context, line: String) {
        val file = getLogFile(context)
        file.appendText(line + "\n")
    }

    private fun rewriteLogs(context: Context) {
        val file = getLogFile(context)
        if (events.isEmpty()) {
            if (file.exists()) {
                file.delete()
            }
        } else {
            val content = events.joinToString("\n", postfix = "\n")
            file.writeText(content)
        }
    }
}
